File: stripe/ios/StripeIntegration.swift
// DreamCobots — Stripe iOS (Swift) Integration
//
// Provides Stripe Checkout and PaymentSheet integration for iOS apps.
//
// Setup:
//   1. Add pod 'StripePaymentSheet' to your Podfile and run `pod install`
//   2. Set your Publishable Key in AppDelegate or SwiftUI App entry point
//   3. Implement a backend endpoint that creates a PaymentIntent and
//      returns { clientSecret: "..." }
//
// This file shows the SwiftUI-based PaymentSheet integration.

import SwiftUI
import StripePaymentSheet

// MARK: - Payment Sheet View Model

@MainActor
final class PaymentViewModel: ObservableObject {

    @Published var paymentSheet: PaymentSheet?
    @Published var paymentResult: PaymentSheetResult?
    @Published var isLoading = false

    private let backendURL: URL

    init(backendURL: URL = URL(string: "https://your-backend.example.com")!) {
        self.backendURL = backendURL
    }

    /// Fetch a PaymentIntent client secret from your backend, then prepare the PaymentSheet.
    func preparePaymentSheet(amountInCents: Int, currency: String) async {
        isLoading = true
        defer { isLoading = false }

        guard let url = URL(string: "\(backendURL)/create-payment-intent") else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "amount": amountInCents,
            "currency": currency
        ])

        guard let (data, _) = try? await URLSession.shared.data(for: request),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let clientSecret = json["clientSecret"] as? String,
              let publishableKey = json["publishableKey"] as? String else {
            return
        }

        STPAPIClient.shared.publishableKey = publishableKey

        var configuration = PaymentSheet.Configuration()
        configuration.merchantDisplayName = "DreamCobots"
        configuration.allowsDelayedPaymentMethods = true

        paymentSheet = PaymentSheet(paymentIntentClientSecret: clientSecret,
                                   configuration: configuration)
    }

    /// Handle the result from the presented PaymentSheet.
    func handlePaymentResult(_ result: PaymentSheetResult) {
        paymentResult = result
        switch result {
        case .completed:
            print("Payment succeeded")
        case .canceled:
            print("Payment canceled by user")
        case .failed(let error):
            print("Payment failed: \(error.localizedDescription)")
        }
    }
}

// MARK: - Payment View

struct PaymentView: View {

    @StateObject private var viewModel = PaymentViewModel()

    var body: some View {
        VStack(spacing: 20) {
            Text("DreamCobots Payment")
                .font(.title2)
                .bold()

            if viewModel.isLoading {
                ProgressView("Preparing payment…")
            } else if let sheet = viewModel.paymentSheet {
                PaymentSheet.PaymentButton(
                    paymentSheet: sheet,
                    onCompletion: viewModel.handlePaymentResult
                ) {
                    Label("Pay Now", systemImage: "creditcard")
                        .frame(maxWidth: .infinity)
                        .padding()
                        .background(Color.accentColor)
                        .foregroundColor(.white)
                        .cornerRadius(10)
                }
            } else {
                Button("Prepare Payment") {
                    Task { await viewModel.preparePaymentSheet(amountInCents: 2500, currency: "usd") }
                }
                .buttonStyle(.borderedProminent)
            }

            if let result = viewModel.paymentResult {
                switch result {
                case .completed:
                    Label("Payment successful!", systemImage: "checkmark.circle.fill")
                        .foregroundColor(.green)
                case .canceled:
                    Text("Payment canceled.")
                        .foregroundColor(.secondary)
                case .failed(let error):
                    Label("Payment failed: \(error.localizedDescription)",
                          systemImage: "xmark.circle.fill")
                        .foregroundColor(.red)
                }
            }
        }
        .padding()
    }
}

// MARK: - App Entry Point (example)

@main
struct DreamCobotsApp: App {
    var body: some Scene {
        WindowGroup {
            PaymentView()
        }
    }
}

File: stripe/android/StripeIntegration.kt
// DreamCobots — Stripe Android (Kotlin) Integration
//
// Provides PaymentSheet integration for Android apps.
//
// Setup:
//   1. Add `implementation("com.stripe:stripe-android:20.+")` to app/build.gradle
//   2. Implement a backend endpoint that creates a PaymentIntent and
//      returns { clientSecret: "...", publishableKey: "..." }
//   3. Initialize PaymentSheet in your Activity or Fragment as shown below.

package com.dreamcobots.stripe

import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.lifecycle.lifecycleScope
import com.stripe.android.PaymentConfiguration
import com.stripe.android.paymentsheet.PaymentSheet
import com.stripe.android.paymentsheet.PaymentSheetResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL

private const val TAG = "DreamCobotsStripe"
private const val BACKEND_URL = "https://your-backend.example.com"

// ---------------------------------------------------------------------------
// Main Activity
// ---------------------------------------------------------------------------

class MainActivity : ComponentActivity() {

    private lateinit var paymentSheet: PaymentSheet
    private var clientSecret: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        paymentSheet = PaymentSheet(this, ::onPaymentSheetResult)

        setContent {
            MaterialTheme {
                PaymentScreen(
                    onPrepare = { preparePayment(amountInCents = 2500, currency = "usd") },
                    onPay = { clientSecret?.let { presentPaymentSheet(it) } }
                )
            }
        }
    }

    // -----------------------------------------------------------------------
    // Fetch PaymentIntent from backend
    // -----------------------------------------------------------------------

    private fun preparePayment(amountInCents: Int, currency: String) {
        lifecycleScope.launch {
            val result = withContext(Dispatchers.IO) {
                runCatching {
                    val url = URL("$BACKEND_URL/create-payment-intent")
                    val connection = url.openConnection() as java.net.HttpURLConnection
                    connection.requestMethod = "POST"
                    connection.setRequestProperty("Content-Type", "application/json")
                    connection.doOutput = true

                    val body = """{"amount":$amountInCents,"currency":"$currency"}"""
                    connection.outputStream.use { it.write(body.toByteArray()) }

                    val response = connection.inputStream.bufferedReader().readText()
                    JSONObject(response)
                }
            }

            result.onSuccess { json ->
                val publishableKey = json.getString("publishableKey")
                clientSecret = json.getString("clientSecret")
                PaymentConfiguration.init(applicationContext, publishableKey)
                Log.d(TAG, "PaymentIntent prepared, client secret received")
            }.onFailure { error ->
                Log.e(TAG, "Failed to prepare payment: ${error.message}")
            }
        }
    }

    // -----------------------------------------------------------------------
    // Present PaymentSheet
    // -----------------------------------------------------------------------

    private fun presentPaymentSheet(secret: String) {
        val configuration = PaymentSheet.Configuration(
            merchantDisplayName = "DreamCobots",
            allowsDelayedPaymentMethods = true
        )
        paymentSheet.presentWithPaymentIntent(secret, configuration)
    }

    // -----------------------------------------------------------------------
    // Handle result
    // -----------------------------------------------------------------------

    private fun onPaymentSheetResult(result: PaymentSheetResult) {
        when (result) {
            is PaymentSheetResult.Completed -> Log.i(TAG, "Payment succeeded!")
            is PaymentSheetResult.Canceled -> Log.i(TAG, "Payment canceled by user")
            is PaymentSheetResult.Failed -> Log.e(TAG, "Payment failed: ${result.error.message}")
        }
    }
}

// ---------------------------------------------------------------------------
// Compose UI
// ---------------------------------------------------------------------------

@Composable
fun PaymentScreen(onPrepare: () -> Unit, onPay: () -> Unit) {
    var prepared by remember { mutableStateOf(false) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Text("DreamCobots Payment", style = MaterialTheme.typography.headlineSmall)
        Spacer(Modifier.height(24.dp))

        if (!prepared) {
            Button(onClick = {
                onPrepare()
                prepared = true
            }) {
                Text("Prepare Payment")
            }
        } else {
            Button(onClick = onPay) {
                Text("Pay Now")
            }
        }
    }
}
